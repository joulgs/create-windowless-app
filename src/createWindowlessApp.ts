import * as commander from "commander";
import { Command } from "commander";
import chalk from "chalk";
import * as envinfo from "envinfo";
import * as path from "path";
import * as fs from "fs-extra";
import validateProjectName from 'validate-npm-package-name';
import * as os from "os";
import spawn from "cross-spawn";

const packageJson = require('../package.json');

const consts = require('../resources/consts.json');

// These files should be allowed to remain on a failed install, but then silently removed during the next create.
const errorLogFilePatterns = consts.errorLogFilePatterns;

export function createWindowlessApp() {
    let projectName: string = undefined;

    const program: Command = new commander.Command(packageJson.name)
        .version(packageJson.version)
        .arguments('<project-directory>')
        .usage(`${ chalk.green('<project-directory>') } [options]`)
        .action(name => {
            projectName = name;
        })
        .option('--verbose', 'print additional logs')
        .option('--info', 'print environment debug info')
        .option('--typescript')
        .allowUnknownOption()
        .on('--help', () => {
            console.log(`    Only ${ chalk.green('<project-directory>') } is required.`);
            console.log();
            console.log(`    If you have any problems, do not hesitate to file an issue:`);
            console.log(`      ${ chalk.cyan('https://github.com/yoavain/create-windowless-app/issues/new') }`);
            console.log();
        })
        .parse(process.argv);

    if (program.info) {
        console.log(chalk.bold('\nEnvironment Info:'));
        return envinfo
            .run({
                    System: ['OS', 'CPU'],
                    Binaries: ['Node', 'npm'],
                    npmPackages: [...consts.dependencies, ...consts.devDependencies, ...consts.tsDevDependencies],
                    npmGlobalPackages: ['create-windowless-app'],
                },
                {
                    duplicates: true,
                    showNotFound: true,
                }
            )
            .then(console.log);
    }

    if (typeof projectName === 'undefined') {
        console.error('Please specify the project directory:');
        console.log(`  ${ chalk.cyan(program.name()) } ${ chalk.green('<project-directory>') }`);
        console.log();
        console.log('For example:');
        console.log(`  ${ chalk.cyan(program.name()) } ${ chalk.green('my-windowless-app') }`);
        console.log();
        console.log(`Run ${ chalk.cyan(`${ program.name() } --help`) } to see all options.`);
        process.exit(1);
    }

    function printValidationResults(results) {
        if (typeof results !== 'undefined') {
            results.forEach(error => {
                console.error(chalk.red(`  *  ${ error }`));
            });
        }
    }

    const hiddenProgram = new commander.Command()
        .option('--internal-testing-template <path-to-template>', `(internal usage only, DO NOT RELY ON THIS) use a non-standard application template`)
        .parse(process.argv);

    createApp(projectName, program.verbose, program.typescript);
}

function createApp(name: string, verbose: boolean, useTypescript: boolean) {
    const root = path.resolve(name);
    const appName = path.basename(root);

    checkAppName(appName);
    fs.ensureDirSync(name);
    if (!isSafeToCreateProjectIn(root, name)) {
        process.exit(1);
    }

    console.log(`Creating a new windowless app in ${chalk.green(root)}.`);
    console.log();

    const packageJson = {
        name: appName,
        version: '0.1.0',
        private: true,
    };
    fs.writeFileSync(path.join(root, 'package.json'), JSON.stringify(packageJson, null, 2) + os.EOL);

    const originalDirectory = process.cwd();
    process.chdir(root);
    if (!checkThatNpmCanReadCwd()) {
        process.exit(1);
    }

    run(root, appName, verbose, originalDirectory, useTypescript);
}

function run(root: string, appName: string, verbose: boolean, originalDirectory: string, useTypescript: boolean): Promise<void> {
    const dependencies = [...consts.dependencies];
    const devDependencies = [...consts.devDependencies];
    if (useTypescript) {
        devDependencies.push(...consts.tsDevDependencies);
    }

    return install(root, dependencies, verbose, false)
        .then(() => {
            return install(root, devDependencies, verbose, true);
        })
        .then(() => console.log("Done"));
}

function install(root: string, dependencies: string[], verbose: boolean, isDev: boolean): Promise<void> {
    return new Promise((resolve, reject) => {
        const command = 'npm';
        let args = ['install', isDev ? '--save-dev' : '--save', '--save-exact', '--loglevel', 'error'].concat(dependencies);
        if (verbose) {
            args.push('--verbose');
        }

        const child = spawn(command, args, { stdio: 'inherit' });
        child.on('close', code => {
            if (code !== 0) {
                reject({
                    command: `${command} ${args.join(' ')}`,
                });
                return;
            }
            resolve();
        });
    });
}

function checkAppName(appName) {
    const validationResult = validateProjectName(appName);
    if (!validationResult.validForNewPackages) {
        console.error(`Could not create a project called ${ chalk.red(`"${ appName }"`) } because of npm naming restrictions:`);
        printValidationResults(validationResult.errors);
        printValidationResults(validationResult.warnings);
        process.exit(1);
    }

    const dependencies = [...consts.dependencies, ...consts.devDependencies].sort();
    if (dependencies.indexOf(appName) >= 0) {
        console.error(chalk.red(`We cannot create a project called ${ chalk.green(appName) } because a dependency with the same name exists.\n` +
            `Due to the way npm works, the following names are not allowed:\n\n`) +
            chalk.cyan(dependencies.map(depName => `  ${ depName }`).join('\n')) +
            chalk.red('\n\nPlease choose a different project name.')
        );
        process.exit(1);
    }
}

function printValidationResults(results) {
    if (typeof results !== 'undefined') {
        results.forEach(error => {
            console.error(chalk.red(`  *  ${ error }`));
        });
    }
}

// If project only contains files generated by GH, itâ€™s safe.
// Also, if project contains remnant error logs from a previous installation, lets remove them now.
// We also special case IJ-based products .idea because it integrates with CRA:
// https://github.com/facebook/create-react-app/pull/368#issuecomment-243446094
function isSafeToCreateProjectIn(root, name) {
    const validFiles: string[] = consts.validFiles;
    console.log();

    const conflicts = fs
        .readdirSync(root)
        .filter(file => !validFiles.includes(file))
        // IntelliJ IDEA creates module files before CRA is launched
        .filter(file => !/\.iml$/.test(file))
        // Don't treat log files from previous installation as conflicts
        .filter(file => !errorLogFilePatterns.some(pattern => file.indexOf(pattern) === 0));

    if (conflicts.length > 0) {
        console.log(`The directory ${ chalk.green(name) } contains files that could conflict:`);
        console.log();
        for (const file of conflicts) {
            console.log(`  ${ file }`);
        }
        console.log();
        console.log('Either try using a new directory name, or remove the files listed above.');

        return false;
    }

    // Remove any remnant files from a previous installation
    const currentFiles = fs.readdirSync(path.join(root));
    currentFiles.forEach(file => {
        errorLogFilePatterns.forEach(errorLogFilePattern => {
            // This will catch `npm-debug.log*` files
            if (file.indexOf(errorLogFilePattern) === 0) {
                fs.removeSync(path.join(root, file));
            }
        });
    });
    return true;
}

function checkThatNpmCanReadCwd() {
    const cwd = process.cwd();
    let childOutput = null;
    try {
        // Note: intentionally using spawn over exec since
        // the problem doesn't reproduce otherwise.
        // `npm config list` is the only reliable way I could find
        // to reproduce the wrong path. Just printing process.cwd()
        // in a Node process was not enough.
        childOutput = spawn.sync('npm', ['config', 'list']).output.join('');
    } catch (err) {
        // Something went wrong spawning node.
        // Not great, but it means we can't do this check.
        // We might fail later on, but let's continue.
        return true;
    }
    if (typeof childOutput !== 'string') {
        return true;
    }
    const lines = childOutput.split('\n');
    // `npm config list` output includes the following line:
    // "; cwd = C:\path\to\current\dir" (unquoted)
    // I couldn't find an easier way to get it.
    const prefix = '; cwd = ';
    const line = lines.find(line => line.indexOf(prefix) === 0);
    if (typeof line !== 'string') {
        // Fail gracefully. They could remove it.
        return true;
    }
    const npmCWD = line.substring(prefix.length);
    if (npmCWD === cwd) {
        return true;
    }
    console.error(
        chalk.red(`Could not start an npm process in the right directory.\n\n` +
            `The current directory is: ${chalk.bold(cwd)}\n` +
            `However, a newly started npm process runs in: ${chalk.bold(npmCWD)}\n\n` +
            `This is probably caused by a misconfigured system terminal shell.`
        )
    );
    if (process.platform === 'win32') {
        console.error(chalk.red(`On Windows, this can usually be fixed by running:\n\n`) +
            `  ${chalk.cyan('reg')} delete "HKCU\\Software\\Microsoft\\Command Processor" /v AutoRun /f\n` +
            `  ${chalk.cyan('reg')} delete "HKLM\\Software\\Microsoft\\Command Processor" /v AutoRun /f\n\n` +
            chalk.red(`Try to run the above two lines in the terminal.\n`) +
            chalk.red(`To learn more about this problem, read: https://blogs.msdn.microsoft.com/oldnewthing/20071121-00/?p=24433/`)
        );
    }
    return false;
}