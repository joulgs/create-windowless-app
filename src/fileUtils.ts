import * as fs from "fs-extra";
import path from "path";
import os from "os";
import { PathLike } from "fs";

export const readFile = (fileName: string): string => {
    return fs.readFileSync(fileName, "utf8");
};

export const readJsonFile = (jsonFileName: string): any => {
    return JSON.parse(readFile(jsonFileName));
};

export const readResource = (resourceRelativePath): string => {
    return readFile(path.resolve(__dirname, resourceRelativePath));
};

export const readJsonResource = (resourceRelativePath): any => {
    return JSON.parse(readResource(resourceRelativePath));
};

export const writeJson = (fileName: string, object): void => {
    fs.writeFileSync(fileName, JSON.stringify(object, null, 2).replace(/\n/g, os.EOL) + os.EOL);
};

export const writeFile = (fileName: string, data: string): void => {
    fs.writeFileSync(fileName, data.replace(/\r/g, "").replace(/\n/g, os.EOL));
};

export const copyFile = (source: PathLike, destination: PathLike): void => {
    fs.copyFileSync(source, destination);
};
