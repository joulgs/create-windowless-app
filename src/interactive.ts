import inquirer from "inquirer";
import type { ProgramConfig } from "./cliParser";
import { validateNodeVersion, validateProjectNameInput } from "./validation";

export const interactiveMode = (): Promise<ProgramConfig> => {
    return inquirer.prompt([
        {
            type: "input",
            message: "Project Name:",
            name: "projectName",
            validate: validateProjectNameInput
        },
        {
            type: "input",
            message: "Icon:",
            name: "icon"
        },
        {
            type: "confirm",
            message: "TypeScript:",
            name: "typescript",
            default: true
        },
        {
            type: "confirm",
            message: "Install Husky:",
            name: "husky",
            default: true
        },
        {
            type: "confirm",
            message: "Skip NPM Install:",
            name: "skipInstall",
            default: false
        },
        {
            type: "input",
            message: "Node Version:",
            name: "nodeVersion",
            validate: validateNodeVersion
        },
        {
            type: "confirm",
            message: "Verbose:",
            name: "verbose",
            default: false
        }
    ]);
};