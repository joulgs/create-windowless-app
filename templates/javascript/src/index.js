const path = require("path");
const winston = require('winston');
const { WindowsToaster } = require('node-notifier');

// Notifier init
const argv = process.argv;
const snoreToastPath = argv[0] === "node.exe" ? null : path.join(argv[0], "../", "SnoreToast.exe");
const notifier = new WindowsToaster({withFallback: !!snoreToastPath, customPath: snoreToastPath});

// Logger init
const { combine, timestamp, printf, label } = winston.format;
const transports = {
    file: new winston.transports.File({ filename: "app.log" })
};
transports.file.level = "debug";
const logger = winston.createLogger({
    level: 'debug',
    format: combine(
        label({ label: '[my-label]' }),
        timestamp({ format: 'YYYY-MM-DD HH:mm:ss.SSS' }),
        printf(info => `${ info.timestamp } [${ info.level.toUpperCase() }] ${ info.message }`)
    ),
    transports: [transports.file]
});


// Log message
logger.log("info", "Hello World");

// Notify
notifier.notify({title: 'APP NAME',  message: 'Hello World' });
