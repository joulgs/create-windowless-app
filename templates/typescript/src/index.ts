import * as path from "path";
import * as winston from 'winston';
import { WindowsToaster } from 'node-notifier';

// App Name
const AppName = "<APPNAME>";

// Args (ignore exe + js)
const argv: string[] = process.argv.slice(2);

// Notifier init
const snoreToastPath = argv[0].endsWith(".exe")  ? path.join(argv[0], "../", "SnoreToast.exe") : null;
const notifier = new WindowsToaster({withFallback: !!snoreToastPath, customPath: snoreToastPath});

// Logger init
const {combine, timestamp, printf, label} = winston.format;
const transports = {
    file: new winston.transports.File({filename: "app.log"})
};
transports.file.level = "debug";
const logger: winston.Logger = winston.createLogger({
    level: 'debug',
    format: combine(
        label({label: '[my-label]'}),
        timestamp({format: 'YYYY-MM-DD HH:mm:ss.SSS'}),
        printf(info => `${info.timestamp} [${info.level.toUpperCase()}] ${info.message}`)
    ),
    transports: [transports.file]
});


// Log message
logger.log("info", `"${AppName}" started with ${argv ? argv.join("; ") : "no args"}`);

// Notify
notifier.notify({title: `${AppName}`,  message: 'Hello World' });
