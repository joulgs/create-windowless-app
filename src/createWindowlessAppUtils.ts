import validateProjectName from "validate-npm-package-name";
import chalk from "chalk";
import path from "path";
import { readJsonFile, writeJson } from "./fileUtils";
import consts from "./consts";

export const PACKAGE_JSON_FILENAME = "package.json";

const printValidationResults = (results) => {
    if (typeof results !== "undefined") {
        results.forEach((error) => {
            console.error(chalk.red(`  *  ${error}`));
        });
    }
};

export const checkAppName = (appName) => {
    const validationResult = validateProjectName(appName);
    if (!validationResult.validForNewPackages) {
        console.error(`Could not create a project called ${chalk.red(`"${appName}"`)} because of npm naming restrictions:`);
        printValidationResults(validationResult.errors);
        printValidationResults(validationResult.warnings);
        process.exit(1);
    }

    const dependencies = [...consts.dependencies, ...consts.devDependencies].sort();
    if (dependencies.indexOf(appName) >= 0) {
        console.error(
            chalk.red(
                `We cannot create a project called ${chalk.green(appName)} because a dependency with the same name exists.\n` + "Due to the way npm works, the following names are not allowed:\n\n"
            ) +
            chalk.cyan(dependencies.map((depName) => `  ${depName}`).join("\n")) +
            chalk.red("\n\nPlease choose a different project name.")
        );
        process.exit(1);
    }
};

export const getNexeCommand = (appName: string, nodeVersion: string) => {
    return `nexe -t ${nodeVersion} -o dist/${appName}.exe`;
};

export const mergeIntoPackageJson = (root: string, field: string, data: any) => {
    const packageJsonPath = path.resolve(root, PACKAGE_JSON_FILENAME);
    const packageJson = readJsonFile(packageJsonPath);
    if (Array.isArray(data)) {
        const list = (packageJson[field] || []).concat(data).reduce((acc, cur) => {
            acc[cur] = cur;
            return acc;
        }, {});
        packageJson[field] = Object.keys(list);
    }
    else {
        packageJson[field] = Object.assign(packageJson[field] || {}, data);
    }
    writeJson(packageJsonPath, packageJson);
};

export const replaceAppNamePlaceholder = (str: string, appName: string): string => {
    return str.replace(/<APPNAME>/g, `${appName}`);
};
