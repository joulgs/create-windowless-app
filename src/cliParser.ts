import commander from "commander";
import chalk from "chalk";
import { PACKAGE_JSON_FILENAME } from "~src/createWindowlessAppUtils";
import * as fs from "fs-extra";
import inquirer from "inquirer";
import type { InvalidNames, LegacyNames, ValidNames } from "validate-npm-package-name";
import validateProjectName from "validate-npm-package-name";
import semver from "semver";

const packageJson = require(`../${PACKAGE_JSON_FILENAME}`);

export type ParsedCommand = {
    projectName: string,
    command: commander.Command
}

export type ProgramConfig = {
    projectName: string;
    icon?: string;
    typescript: boolean;
    husky: boolean;
    skipInstall: boolean;
    nodeVersion: string;
    verbose: boolean;
};

function interactiveMode(): Promise<ProgramConfig> {
    return inquirer.prompt([
        {
            type: "input",
            message: "Project Name:",
            name: "projectName",
            validate: (value) => {
                const result: ValidNames | InvalidNames | LegacyNames = validateProjectName(value);
                return result.validForNewPackages || ((result as InvalidNames)?.errors?.[0]) || ((result as LegacyNames)?.warnings?.[0]) || "Invalid project name";
            }
        },
        {
            type: "input",
            message: "Icon:",
            name: "icon"
        },
        {
            type: "confirm",
            message: "TypeScript:",
            name: "typescript",
            default: true
        },
        {
            type: "confirm",
            message: "Install Husky:",
            name: "husky",
            default: true
        },
        {
            type: "confirm",
            message: "Skip NPM Install:",
            name: "skipInstall",
            default: false
        },
        {
            type: "input",
            message: "Node Version:",
            name: "nodeVersion",
            validate: (value) => !value || !!semver.valid(value) || "Invalid node version"
        },
        {
            type: "confirm",
            message: "Verbose:",
            name: "verbose",
            default: false
        }
    ]);
}

const validateInput = (programConfig: ProgramConfig, command: commander.Command): ProgramConfig => {
    const programConfigResult = { ...programConfig };
    if (!programConfigResult.projectName || typeof programConfigResult.projectName === "undefined") {
        console.error(`${chalk.red("Missing project name")}`);
        console.log();
        command.outputHelp();
        process.exit(1);
    }

    if (programConfigResult.icon && !fs.pathExistsSync(programConfigResult.icon)) {
        console.log(`Cannot find icon in ${chalk.red(programConfigResult.icon)}. Switching to ${chalk.green("default")} icon.`);
        programConfigResult.icon = undefined;
    }

    return programConfigResult;
};

export const parseCommand = async (argv: string[]): Promise<ProgramConfig> => {
    let projectName: string = undefined;
    const command: commander.Command = new commander.Command(packageJson.name)
        .version(packageJson.version)
        .arguments("<project-directory>")
        .usage(`${chalk.green("<project-directory>")} [options]`)
        .action((name) => {
            projectName = name;
        })
        .option("--verbose", "print additional logs")
        .option("--interactive", "interactive mode")
        .option("--no-typescript", "use javascript rather than typescript")
        .option("--no-husky", "do not install husky pre-commit hook for building launcher")
        .option("--skip-install", "write dependencies to package.json without installing")
        .option("--icon <icon>", "override default launcher icon file")
        .option("--node-version <nodeVersion>", "override node version to bundle")
        .allowUnknownOption()
        .on("--help", () => {
            console.log(`    Only ${chalk.green("<project-directory>")} is required.`);
            console.log();
            console.log("    If you have any problems, do not hesitate to file an issue:");
            console.log(`      ${chalk.cyan("https://github.com/yoavain/create-windowless-app/issues/new")}`);
            console.log();
        })
        .parse(argv);

    let programConfig: ProgramConfig;
    if (command.interactive) {
        programConfig = await interactiveMode();
    }
    else {
        programConfig = {
            projectName,
            verbose: command.verbose,
            typescript: command.typescript,
            husky: command.husky,
            skipInstall: command.skipInstall,
            icon: command.icon,
            nodeVersion: command.nodeVersion
        };
    }

    return validateInput(programConfig, command);
};
