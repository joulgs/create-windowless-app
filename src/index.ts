#!/usr/bin/env node

'use strict';

import { createWindowlessApp } from './createWindowlessApp';
import { checkCscInPath } from "../templates/typescript/src/launcherCompiler";

const currentNodeVersion: string = process.versions.node;
const semver: string[] = currentNodeVersion.split('.');
const major: number = Number(semver[0]);

if (isNaN(major) || major < 8) {
    console.error(`You are running Node ${currentNodeVersion}.\nCreate Windowless App requires Node 8 or higher.\nPlease update your version of Node.`);
    process.exit(1);
}

// Check for csc.exe in %PATH%
checkCscInPath()
    .then(exist => {
        if (exist) {
            // noinspection JSIgnoredPromiseFromCall
            createWindowlessApp();
        }
        else {
            console.error(`You need "csc.exe" (C# compiler) in your path in order to compile the launcher.exe.`);
            process.exit(1);
        }
    });
