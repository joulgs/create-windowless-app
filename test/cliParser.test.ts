import { parseCommand } from "../src/cliParser";
import { v4 as uuid } from "uuid";

describe("Test cliParser", () => {
    afterEach(() => {
        jest.restoreAllMocks();
    });
    
    it("should parse default flags", async () => {
        const sandbox: string = uuid();

        const { projectName, command } = parseCommand(["node.exe", "dummy.ts", sandbox]);

        expect(projectName).toEqual(sandbox);
        expect(command.interactive).toBeUndefined();
        expect(command.verbose).toBeUndefined();
        expect(command.typescript).toEqual(true);
        expect(command.husky).toEqual(true);
        expect(command.skipInstall).toBeUndefined();
        expect(command.icon).toBeUndefined();
        expect(command.nodeVersion).toBeUndefined();
    });

    it("should parse flags: --skip-install", async () => {
        const sandbox: string = uuid();

        const { projectName, command } = parseCommand(["node.exe", "dummy.ts", sandbox, "--skip-install"]);

        expect(projectName).toEqual(sandbox);
        expect(command.interactive).toBeUndefined();
        expect(command.verbose).toBeUndefined();
        expect(command.typescript).toEqual(true);
        expect(command.husky).toEqual(true);
        expect(command.skipInstall).toEqual(true);
        expect(command.icon).toBeUndefined();
        expect(command.nodeVersion).toBeUndefined();
    });

    it("should parse flags: --no-typescript", async () => {
        const sandbox: string = uuid();

        const { projectName, command } = parseCommand(["node.exe", "dummy.ts", sandbox, "--no-typescript"]);

        expect(projectName).toEqual(sandbox);
        expect(command.interactive).toBeUndefined();
        expect(command.verbose).toBeUndefined();
        expect(command.typescript).toEqual(false);
        expect(command.husky).toEqual(true);
        expect(command.skipInstall).toBeUndefined();
        expect(command.icon).toBeUndefined();
        expect(command.nodeVersion).toBeUndefined();
    });

    it("should parse flags: --no-husky", async () => {
        const sandbox: string = uuid();

        const { projectName, command } = parseCommand(["node.exe", "dummy.ts", sandbox, "--no-husky"]);

        expect(projectName).toEqual(sandbox);
        expect(command.interactive).toBeUndefined();
        expect(command.verbose).toBeUndefined();
        expect(command.typescript).toEqual(true);
        expect(command.husky).toEqual(false);
        expect(command.skipInstall).toBeUndefined();
        expect(command.icon).toBeUndefined();
        expect(command.nodeVersion).toBeUndefined();
    });

    it("should parse flags: --verbose", async () => {
        const sandbox: string = uuid();

        const { projectName, command } = parseCommand(["node.exe", "dummy.ts", sandbox, "--verbose"]);

        expect(projectName).toEqual(sandbox);
        expect(command.interactive).toBeUndefined();
        expect(command.verbose).toEqual(true);
        expect(command.typescript).toEqual(true);
        expect(command.husky).toEqual(true);
        expect(command.skipInstall).toBeUndefined();
        expect(command.icon).toBeUndefined();
        expect(command.nodeVersion).toBeUndefined();
    });

    it("should parse flags: --icon", async () => {
        const sandbox: string = uuid();

        const { projectName, command } = parseCommand(["node.exe", "dummy.ts", sandbox, "--icon", "someIcon"]);

        expect(projectName).toEqual(sandbox);
        expect(command.interactive).toBeUndefined();
        expect(command.verbose).toBeUndefined();
        expect(command.typescript).toEqual(true);
        expect(command.husky).toEqual(true);
        expect(command.skipInstall).toBeUndefined();
        expect(command.icon).toEqual("someIcon");
        expect(command.nodeVersion).toBeUndefined();
    });

    it("should parse flags: --node-version", async () => {
        const sandbox: string = uuid();

        const { projectName, command } = parseCommand(["node.exe", "dummy.ts", sandbox, "--node-version", "12.12.0"]);

        expect(projectName).toEqual(sandbox);
        expect(command.interactive).toBeUndefined();
        expect(command.verbose).toBeUndefined();
        expect(command.typescript).toEqual(true);
        expect(command.husky).toEqual(true);
        expect(command.skipInstall).toBeUndefined();
        expect(command.icon).toBeUndefined();
        expect(command.nodeVersion).toEqual("12.12.0");
    });

    it("should print help with flags: --help", async () => {
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        const mockExit = jest.spyOn(process, "exit").mockImplementation(() => {});
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        const mockStdout = jest.spyOn(process.stdout, "write").mockImplementation(() => {});
        const sandbox: string = uuid();

        parseCommand(["node.exe", "dummy.ts", sandbox, "--help"]);

        expect(mockExit).toHaveBeenCalledWith(0);
        expect(mockStdout.mock.calls[0][0]).toMatchSnapshot("help-stdout");
    });

    it("should error on missing project name", async () => {
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        const mockExit = jest.spyOn(process, "exit").mockImplementation(() => {});
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        const mockStdout = jest.spyOn(process.stdout, "write").mockImplementation(() => {});

        const { projectName } = parseCommand(["node.exe", "dummy.ts"]);

        expect(projectName).toBeUndefined();
        expect(mockExit).toHaveBeenCalledWith(1);
        expect(mockStdout.mock.calls[0][0]).toMatchSnapshot("missing-project-name-stdout");
    });
});