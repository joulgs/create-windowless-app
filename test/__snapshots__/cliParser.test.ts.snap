// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cliParser should error on missing project name: missing-project-name-stdout 1`] = `
"  console.error
    jest.js <projectName>
    
    project name
    
    Positionals:
      projectName                                                           [string]
    
    Options:
      -v, --verbose       print additional logs                            [boolean]
      -i, --interactive   interactive mode                                 [boolean]
      -t, --typescript    use typescript                   [boolean] [default: true]
      -h, --husky         install husky pre-commit hook for building launcher
                                                           [boolean] [default: true]
      -s, --skip-install  write dependencies to package.json without installing
                                                                           [boolean]
      -c, --icon          override default launcher icon file               [string]
      -n, --node-version  override node version to bundle                   [string]
          --version       Show version number                              [boolean]
          --help          Show help                                        [boolean]
"
`;

exports[`Test cliParser should print help with flags: --help: help-stdout 1`] = `
"  console.log
    jest.js <projectName>
    
    project name
    
    Positionals:
      projectName  project name                                             [string]
    
    Options:
      -v, --verbose       print additional logs                            [boolean]
      -i, --interactive   interactive mode                                 [boolean]
      -t, --typescript    use typescript                   [boolean] [default: true]
      -h, --husky         install husky pre-commit hook for building launcher
                                                           [boolean] [default: true]
      -s, --skip-install  write dependencies to package.json without installing
                                                                           [boolean]
      -c, --icon          override default launcher icon file               [string]
      -n, --node-version  override node version to bundle                   [string]
          --version       Show version number                              [boolean]
          --help          Show help                                        [boolean]
"
`;
